server/server.go

```diff
@@ -0,0 +1,229 @@
+}
+
+// CreateServer initializes a server with `numClients` clients with a master
+// secret of length `lenMS`, and generate salts with length `lenSalt`.  The
+// number of salts is given by `r = -log2(fpRate)`, where `fpRate` is the
+// desired false positive rate of the system.  `mountPoint` determines where the
+// server files will be stored.
+func CreateServer(numClients, lenMS, lenSalt int, mountPoint string, fpRate float64, numUniqWords uint64) *Server {
+	s := new(Server)
+	masterSecret := make([]byte, lenMS)
+	rand.Read(masterSecret)
+	s.keyHalves = make([][]byte, numClients)
+	s.lenMS = lenMS
+	for i := 0; i < numClients; i++ {
+		h := sha256.New()
+		h.Write([]byte(strconv.Itoa(i)))
+		cksum := h.Sum(nil)
+		s.keyHalves[i] = util.XorBytes(masterSecret, cksum, lenMS)
+	}
+	r := int(math.Ceil(-math.Log2(fpRate)))
+	s.size = uint64(math.Ceil(float64(numUniqWords) * float64(r) / math.Log(2)))
+	s.salts = util.GenerateSalts(r, lenSalt)
+	s.numFiles = 0
+	s.mountPoint = mountPoint
+	s.writeToFile()
+	return s
+}
+
+// CreateServerWithLog behaves the same as `CreateServer`, except for that it
+// also sets the logging parameters for the server.
+func CreateServerWithLog(numClients, lenMS, lenSalt int, mountPoint string, fpRate float64, numUniqWords uint64, latency time.Duration, bandwidth int) *Server {
+	s := CreateServer(numClients, lenMS, lenSalt, mountPoint, fpRate, numUniqWords)
+	s.latency = latency
+	s.bandwidth = bandwidth
+	s.writeToFile()
+	return s
+}
+
+// LoadServer initializes a Server by reading the metadata stored at
+// `mountPoint` and restoring the server status.
+func LoadServer(mountPoint string) *Server {
+	input, err := os.Open(path.Join(mountPoint, "serverMD"))
+	if err != nil {
+		panic("Server metadata not found")
+	}
+	dec := gob.NewDecoder(input)
+
+	s := new(Server)
+	dec.Decode(&s.mountPoint)
+	dec.Decode(&s.numFiles)
+	dec.Decode(&s.salts)
+	dec.Decode(&s.keyHalves)
+	dec.Decode(&s.lenMS)
+	dec.Decode(&s.size)
+	dec.Decode(&s.latency)
+	dec.Decode(&s.bandwidth)
+
+	input.Close()
+
+	return s
+}
+
+// writeToFile serializes the server status and writes the metadata to a file in
+// the server mount point, which can be later loaded by `LoadServer`.
+func (s *Server) writeToFile() {
+	file, _ := os.Create(path.Join(s.mountPoint, "serverMD"))
+	enc := gob.NewEncoder(file)
+	enc.Encode(s.mountPoint)
+	enc.Encode(s.numFiles)
+	enc.Encode(s.salts)
+	enc.Encode(s.keyHalves)
+	enc.Encode(s.lenMS)
+	enc.Encode(s.size)
+	enc.Encode(s.latency)
+	enc.Encode(s.bandwidth)
+
+	file.Close()
+}
+
+// AddFile adds a file with `content` to the server with the document ID equal
+// to the number of files currently in the server and updates the count.
+// Returns the document ID.
+func (s *Server) AddFile(content []byte) int {
+	logger.AddTime(s.latency * 2)
+	logger.AddTime(time.Millisecond * time.Duration(float64(len(content))*1.5*8*1000/float64(s.bandwidth)))
```

Comment:
Can you document why "1.5" is needed here (to account for encryption, I assume).

Could you use `time.Second` instead of `time.Millisecond`, and then drop the `*1000`?  Or maybe you need to do it this way because otherwise rounding causes problems?